File: solutions/java/isbn-verifier/1/src/main/java/IsbnVerifier.java
class IsbnVerifier {

    boolean isValid(String isbn) {

        var cleaned = isbn.replace("-", "");
        if (cleaned.length() != 10) {
            return false;
        }

        int sum = 0;
        for (int i = 0; i < 10; i++) {

        var c = cleaned.charAt(i);
        if (Character.isDigit(c)) {
        sum += Character.getNumericValue(c) * (10 - i);
        continue;
        }

        if (i < 9 || c != 'X') {
        return false;
        }

        sum += 10;
        }


        return sum % 11 == 0;
    }

}

File: solutions/kotlin/isbn-verifier/1/src/main/kotlin/IsbnVerifier.kt
class IsbnVerifier {

    fun isValid(number: String): Boolean {
        // Normalize by removing dashes; other characters must be validated explicitly
        val normalized = number.replace("-", "")
        if (normalized.length != ISBN_LENGTH) return false

        val body = normalized.dropLast(1)
        if (!body.all { it.isDigit() }) return false

        val checkValue = when (val check = normalized.last()) {
            in '0'..'9' -> check.digitToInt()
            'X' -> 10
            else -> return false
        }

        val sum = body
            .withIndex()
            .sumOf { (index, c) ->
                c.digitToInt() * (10 - index)
            } + checkValue

        return sum % 11 == 0
    }

    private companion object {
        const val ISBN_LENGTH = 10
    }
}
